File: rotate_array_kplaces.py
"""
// Time Complexity : O(n) n is length of shorter array
// Space Complexity : O(1)
// Did this code successfully run on Leetcode : Yes
// Any problem you faced while coding this : 

// Your code here along with comments explaining your approach
Algorithm Explanation
Reverse the array
Swap the elements from 0 to k-1
Swap the elements from k to end
"""

class Solution:
    def rotate(self, nums: List[int], k: int) -> None:
        def swap(i,j):
            while i < j:
                nums[i],nums[j] = nums[j],nums[i]
                i+=1
                j-=1
        
        """
        Do not return anything, modify nums in-place instead.
        """
        nums.reverse()
        #swap elements from 0 to k-1
        swap(0,k-1)
        
        #swap the elements from k to end
        swap(k,len(nums)-1)

File: RotateArrayKPlaces.cs

// Time Complexity : O(n)
// Space Complexity : O(1)
// Did this code successfully run on Leetcode :
// Any problem you faced while coding this :


// Your code here along with comments explaining your approach
	// 1) Reverse array in place for n-k and 0 to k
	// 2) Then reverse whole array

public void Rotate(int[] nums, int k) {
        
        if(nums == null || nums.Length == 0)
            return;
        
        int n = nums.Length;
        
        if(k > n)
            k = k % n;
        reverse(nums, n-k, n-1);
        reverse(nums, 0, n-k-1);
        reverse(nums, 0, n-1);
        
        //OR
        // reverse(nums, 0, n-1);
        // reverse(nums, 0, k-1);
        // reverse(nums, k, n-1);
    }
    
    private void reverse(int[] nums, int l, int r)
    {
        while(l <= r)
        {
            int temp = nums[l];
            nums[l] = nums[r];
            nums[r] = temp;
            
            l++;
            r--;
        }
    }
}